
# Applying AnchorWave to Address Plant Genome Alignment

The process uses function


## Input data
```
wget https://ftp.ensemblgenomes.org/pub/plants/release-55/gff3/zea_mays/Zea_mays.Zm-B73-REFERENCE-NAM-5.0.55.gff3.gz
wget https://ftp.ensemblgenomes.org/pub/plants/release-55/fasta/zea_mays/dna/Zea_mays.Zm-B73-REFERENCE-NAM-5.0.dna.toplevel.fa.gz
wget https://download.maizegdb.org/Zm-Mo17-REFERENCE-CAU-1.0/Zm-Mo17-REFERENCE-CAU-1.0.fa.gz
gunzip *.gz
#change format 
sed -i 's/>chr/>/g' Zm-Mo17-REFERENCE-CAU-1.0.fa
```


## Major Steps


### Extract CDS and lift over reference and query genome

For AnchorWave, “gff2seq” function is used for extracting CDS sequences, ‘r’, ‘i’ and ‘o’ represents input reference FASTA, GFF(3) file and output files. For minimap2, “x splice” function represents long-read splice alignment, ‘t’ represents the number of threads, ‘k’ represents k-mer size, ‘p’ represents min score ratio and ‘a’ represents that output file is SAM format. Other parameters are default. 

```
#extract cds from maize ref 
anchorwave gff2seq -i Zea_mays.Zm-B73-REFERENCE-NAM-5.0.55.gff3 -r Zea_mays.Zm-B73-REFERENCE-NAM-5.0.dna.toplevel.fa  -o cds.fa
# lift over re
minimap2 -x splice -t 10 -k 12 -a -p 0.4 -N 20 Zm-Mo17-REFERENCE-CAU-1.0.fa cds.fa > cds.sam
# lift over quer
minimap2 -x splice -t 10 -k 12 -a -p 0.4 -N 20 Zea_mays.Zm-B73-REFERENCE-NAM-5.0.dna.toplevel.fa cds.fa > ref.sam

```


### Visualization the result of collinearity

```
#convert to the SAM file
perl alignmentToDotplot.pl Zea_mays.AGPv4.34.gff3 cds.sam > maizecds.tab
```
and then use R to plot Figure1
```
#Use this piece of R to draw a dotplot.
library(ggplot2)
#Transform Coordinates using a function.
changetoM <- function ( position ){
  position=position/1000000;
  paste(position, "M", sep="")
}
#Read gene position, belong to which chromosome and so on
data =read.table("maizecds.tab")
#Select all euchromosomes as factor, modify the chromosomes number manually.
data = data[which(data$V1 %in% c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10" )),]
data = data[which(data$V3 %in% c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")),]
data$V1 = factor(data$V1, levels=c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10" ))
data$V3 = factor(data$V3, levels=c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10" ))
#Using ggplot2 to plot a dotplot and beautify it.
png("E:\\plot\\gaoliang\\figure3.png")
ggplot(data=data, aes(x=V4, y=V2))+geom_point(size=0.5, aes(color=V5))+facet_grid(V1~V3, scales="free", space="free" ) +
  labs(x="Mo17", y="B73")+scale_x_continuous(labels=changetoM) + scale_y_continuous(labels=changetoM) +
  theme(axis.line = element_blank(),
        panel.background = element_blank(),
        panel.border = element_rect(fill=NA,color="black", size=0.5, linetype="solid"),
        axis.text.y = element_text( colour = "black"),
        legend.position='none',
        axis.text.x = element_text(angle=300, hjust=0, vjust=1, colour = "black") )
dev.off()

```
![image](https://github.com/Bio-protocol/anchorwave_protocol/blob/master/graphs/figure3.png)

Figure 3. The identified all anchors between the maize B73 and maize Mo17 genome.

### Applying anchorwave to identify the collinearity and WGD

For	AnchorWave, “genoAli” function is used for whole genome alignment without WGD, ‘as’ represents anchor sequence files, ‘i’, ‘r’ and ‘o’ represents input reference GFF(3), FASTA file and output MAF file. ‘a’ and “ar” represents SAM file generated by mapping conserved sequence to reference and query genome, ‘s’ represents query genome file, ‘n’ represents output anchor files. “IV” represents calling inversion. Other parameters are default.

```
anchorwave genoAli -i Zea_mays.Zm-B73-REFERENCE-NAM-5.0.55.gff3 -as cds.fa -r Zea_mays.Zm-B73-REFERENCE-NAM-5.0.dna.toplevel.fa -a cds.sam -ar ref.sam -s Zm-Mo17-REFERENCE-CAU-1.0.fa -n align1.anchors -IV

```
and then use R to plot Figure2

```
changetoM <- function ( position ){
position=position/1000000;
paste(position, "M", sep="")}
library(ggplot2)
data =read.table("align1.anchors", header=TRUE)
#Select all euchromosomes as factor
data = data[which(data$refChr %in% c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")),]
data = data[which(data$queryChr %in% c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")),]
data$refChr = factor(data$refChr, levels=c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10"))
data$queryCh = factor(data$queryChr, levels=c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10"))
png("E:\\R_Scripts\\figure2.png")
ggplot(data=data, aes(x=queryStart, y=referenceStart))+
geom_point(size=0.5, aes(color=strand)) +
facet_grid(refChr~queryChr, scales="free", space="free") +
labs(x="sorghum", y="maize")+scale_x_continuous(labels=changetoM) +
scale_y_continuous(labels=changetoM) +
  theme(axis.line = element_blank(),
        panel.background = element_blank(),
        panel.border = element_rect(fill =NA,color="black", size=0.5, linetype="solid"),
        axis.text.y = element_text( colour = "black"),
        legend.position='none',
        axis.text.x = element_text(angle=300,hjust=0, vjust=0.5, colour = "black") )
dev.off()
```
![image](https://github.com/Bio-protocol/anchorwave_protocol/blob/master/graphs/figure4.png)

Figure 4. The identified collinear anchors between the maize B73 and maize Mo17 genome.
### Base-pair resolved genome alignment
For	AnchorWave, “genoAli” function is used for whole genome alignment without WGD, ‘as’ represents anchor sequence files, ‘i’, ‘r’ and ‘o’ represents input reference GFF(3), FASTA file and output MAF file. ‘a’ and “ar” represents SAM file generated by mapping conserved sequence to reference and query genome, ‘s’ represents query genome file, ‘n’ represents output anchor files. “IV” represents calling inversion. Other parameters are default.

```
anchorwave genoAli -i Zea_mays.Zm-B73-REFERENCE-NAM-5.0.55.gff3 -as cds.fa -r Zea_mays.Zm-B73-REFERENCE-NAM-5.0.dna.toplevel.fa -a cds.sam -ar ref.sam 
-s Zm-Mo17-REFERENCE-CAU-1.0.fa -n anchors.anchors -o b73tomo17.maf -f  b73tomo17.f.maf -IV
```
## Expecteed results

The last step outputs alignment result as MAF formation 

